### PR TITLE
Nested multi select checkbox

### PR DESCRIPTION
**Pull request by Wilker0725**
**Changes**
Description of changes...
* Enhancement indeterminate checkbox

- [delete] branch

- [WARNING] Description of change in `NestedMultiSelect`

- You may refer to this [ticket](https://jira.ship.gov.sg/browse/BOOKINGSG-4616)
- [Figma](https://www.figma.com/file/us90jOpWBahsqK2VAluPvD/Flagship-Design-System?node-id=11728%3A111511&mode=dev)

### CODE REVIEW

File: src/shared/nested-dropdown-list/nested-dropdown-list-helper.ts [hunk 1]
```diff
@@ -176,43 +186,105 @@ export namespace NestedDropdownListHelper {
...
+                );
+
+                const isAllIndeterminate = indeterminateStatus.every(Boolean);
+                const isAllChecked = checkedStatus.every(Boolean);
+                const someInderminate =
+                    indeterminateStatus.filter(Boolean).length;
+                const someChecked = checkedStatus.filter(Boolean).length;
+
+                const indeterminate =
+                    isAllChecked || isAllIndeterminate
```
Comment (qroll) [inline src/shared/nested-dropdown-list/nested-dropdown-list-helper.ts:273]: `isAllIndeterminate` does not work for this case

```
[
    {
        label: "Category 1",
        value: { id: 999, name: "category 1" },
        key: "999",
        subItems: [
            {
                label: "Sub Category A",
                value: {
                    id: 820,
                    name: "Sub category a",
                },
                key: "820",
                subItems: [
                    {
                        label: "Option 1",
                        value: {
                            id: 10001,
                            name: "Option-1",
                        },
                        key: "10001",
                    },
                    {
                        label: "Option 2",
                        value: {
                            id: 10002,
                            name: "Option-2",
                        },
                        key: "10002",
                    },
                    {
                        label: "Option 3",
                        value: {
                            id: 10003,
                            name: "Option-3",
                        },
                        key: "10003",
                    },
                ],
            },
            {
                label: "Sub Category C",
                value: {
                    id: 822,
                    name: "Sub category c",
                },
                key: "822",
                subItems: [
                    {
                        label: "Honey",
                        value: {
                            id: 104,
                            name: "honey",
                        },
                        key: "104",
                    },
                    {
                        label: "Nuts",
                        value: {
                            id: 105,
                            name: "nuts",
                        },
                        key: "105",
                    },
                    {
                        label: "Butter",
                        value: {
                            id: 106,
                            name: "butter",
                        },
                        key: "106",
                    },
                ],
            },
        ],
    },
]
```

File: src/shared/nested-dropdown-list/nested-dropdown-list-helper.ts [hunk 2]
```diff
@@ -176,43 +186,105 @@ export namespace NestedDropdownListHelper {
         return item;
     };
 
-    export const updateCategoryChecked = <V1, V2, V3>(
-        list: FormattedOptionMap<V1, V2, V3>,
+    export const getCategoryChecked = <V1, V2, V3>(
```
Comment (qroll) [inline src/shared/nested-dropdown-list/nested-dropdown-list-helper.ts:189]: what was the reason for switching from recursively updating to looping over the subitems?

the logic feels like it is repeated twice across `update` and the map iteration

File: src/shared/nested-dropdown-list/nested-dropdown-list.tsx [hunk 3]
```diff
@@ -156,6 +156,7 @@ export const NestedDropdownList = <V1, V2, V3>({
 
     const handleSelect = (item: CombinedFormattedOptionProps<V1, V2, V3>) => {
         const { label, keyPath, value } = item;
+        updateSelectedState(keyPath);
```
Comment (qroll) [inline src/shared/nested-dropdown-list/nested-dropdown-list.tsx:159]: does nested dropdown list need to track the selected state? `input-nested-multi-select` would have passed down an updated `selectedKeyPaths` which should trigger the relevant useEffect

File: src/shared/nested-dropdown-list/nested-dropdown-list-helper.ts [hunk 4]
```diff
@@ -43,7 +43,9 @@ export namespace NestedDropdownListHelper {
                     value,
                     expanded: mode === "expand",
                     isSearchTerm: false,
+                    selected: false,
                     checked: false,
+                    indeterminate: undefined,
```
Comment (qroll) [inline src/shared/nested-dropdown-list/nested-dropdown-list-helper.ts:48]: why `undefined` instead of `false`?

**Review by qroll — CHANGES_REQUESTED** [review]

File: src/shared/nested-dropdown-list/nested-dropdown-list-helper.ts (hunk 4 above)
Comment (Wilker0725) [reply src/shared/nested-dropdown-list/nested-dropdown-list-helper.ts:48]: yes. it should `false`, can't recalled why just i put it as `undefined`. 😓

File: src/shared/nested-dropdown-list/nested-dropdown-list.tsx (hunk 3 above)
Comment (Wilker0725) [reply src/shared/nested-dropdown-list/nested-dropdown-list.tsx:159]: yes. my bad. this seems like pointless, the dropdown will closed after selected.

**Review by qroll — APPROVED** [review]

**Pull request merged**